Firebase/CoreDiagnostics/FIRCDLibrary/FIRCoreDiagnostics.m

```diff
@@ -0,0 +1,688 @@
+
+static NSString *const kFIRServiceMLVisionOnDeviceAutoML = @"MLVisionOnDeviceAutoML";
+static NSString *const kFIRServiceMLVisionOnDeviceFace = @"MLVisionOnDeviceFace";
+static NSString *const kFIRServiceMLVisionOnDeviceBarcode = @"MLVisionOnDeviceBarcode";
+static NSString *const kFIRServiceMLVisionOnDeviceText = @"MLVisionOnDeviceText";
+static NSString *const kFIRServiceMLVisionOnDeviceLabel = @"MLVisionOnDeviceLabel";
+static NSString *const kFIRServiceMLVisionOnDeviceObjectDetection =
+    @"MLVisionOnDeviceObjectDetection";
+static NSString *const kFIRServiceMLModelInterpreter = @"MLModelInterpreter";
+
+// Reconcile this with kFIRLoggerInAppMessaging defined in FIRCore+InAppMessaging.m
+// after fiam sdk is released
+static NSString *const kFIRServiceIAM = @"InAppMessaging";
+
+/**
+ * The file name to keep the unique install string.
+ */
+static NSString *const kUniqueInstallFileName = @"FIREBASE_UNIQUE_INSTALL";
+
+/**
+ * @note This should implement the GDTEventDataObject protocol, but can't because of weak-linking.
+ */
+@interface FIRCoreDiagnosticsLog : NSObject
+
+/** The config that will be converted to proto bytes. */
+@property(nonatomic) logs_proto_mobilesdk_ios_ICoreConfiguration config;
+
+@end
+
+@implementation FIRCoreDiagnosticsLog
+
+- (instancetype)initWithConfig:(logs_proto_mobilesdk_ios_ICoreConfiguration)config {
+  self = [super init];
+  if (self) {
+    _config = config;
+  }
+  return self;
+}
+
+// Provided and required by the GDTEventDataObject protocol.
+- (NSData *)transportBytes {
+  pb_ostream_t sizestream = PB_OSTREAM_SIZING;
+
+  // Encode 1 time to determine the size.
+  if (!pb_encode(&sizestream, logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config)) {
+    NSCAssert(NO, @"Error in nanopb encoding for size: %s", PB_GET_ERROR(&sizestream));
+  }
+
+  // Encode a 2nd time to actually get the bytes from it.
+  size_t bufferSize = sizestream.bytes_written;
+  CFMutableDataRef dataRef = CFDataCreateMutable(CFAllocatorGetDefault(), bufferSize);
+  pb_ostream_t ostream = pb_ostream_from_buffer((void *)CFDataGetBytePtr(dataRef), bufferSize);
+  if (!pb_encode(&ostream, logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config)) {
+    NSCAssert(NO, @"Error in nanopb encoding for bytes: %s", PB_GET_ERROR(&ostream));
+  }
+  CFDataSetLength(dataRef, ostream.bytes_written);
+
+  return CFBridgingRelease(dataRef);
+}
+
+- (void)dealloc {
+  pb_release(logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config);
+}
+
+@end
+
+/** This class produces a protobuf containing diagnostics and usage data to be logged. */
+@interface FIRCoreDiagnostics : NSObject <FIRCoreDiagnosticsInterop>
+
+/** The queue on which all diagnostics collection will occur. */
+@property(nonnull, nonatomic) dispatch_queue_t diagnosticsQueue;
+
+/** The transport object used to send data. */
+@property(nonnull, nonatomic) GDTTransport *transport;
+
+@end
+
+@implementation FIRCoreDiagnostics
```

Comment:
Would it be possible to add basic tests for the class e.g. to verify populated proto, etc.

Reply:
Missed this--tests were in a subsequent PR (you've now seen/added to them).